File: content/snippets/cpp/interview_lcci/03/03_stack_of_plates.cpp
#include <iostream>
#include <vector>

class StackOfPlates {
public:
    explicit StackOfPlates(int cap) : capacity(cap) {}

    void push(int val) {
        if (capacity <= 0) return;
        if (stacks.empty() || stacks.back().size() == static_cast<size_t>(capacity)) {
            stacks.push_back({});
        }
        stacks.back().push_back(val);
    }

    int pop() {
        return popAt(static_cast<int>(stacks.size()) - 1);
    }

    int popAt(int index) {
        if (index < 0 || index >= static_cast<int>(stacks.size()) || stacks[index].empty()) {
            return -1;
        }
        int val = stacks[index].back();
        stacks[index].pop_back();
        if (stacks[index].empty()) {
            stacks.erase(stacks.begin() + index);
        }
        return val;
    }

private:
    int capacity;
    std::vector<std::vector<int>> stacks;
};

int main() {
    StackOfPlates s(2);
    s.push(1);
    s.push(2);
    s.push(3);
    s.push(4);
    std::cout << s.popAt(0) << '\n';
    std::cout << s.pop() << '\n';
    std::cout << s.pop() << '\n';
    std::cout << s.pop() << '\n';
    return 0;
}

File: content/snippets/python/interview_lcci/03/03_stack_of_plates.py
class StackOfPlates:
    def __init__(self, cap: int) -> None:
        self.cap = cap
        self.stacks: list[list[int]] = []

    def push(self, val: int) -> None:
        if self.cap <= 0:
            return
        if not self.stacks or len(self.stacks[-1]) == self.cap:
            self.stacks.append([])
        self.stacks[-1].append(val)

    def pop(self) -> int:
        return self.pop_at(len(self.stacks) - 1)

    def pop_at(self, index: int) -> int:
        if index < 0 or index >= len(self.stacks) or not self.stacks[index]:
            return -1
        val = self.stacks[index].pop()
        if not self.stacks[index]:
            self.stacks.pop(index)
        return val


def main() -> None:
    s = StackOfPlates(2)
    s.push(1)
    s.push(2)
    s.push(3)
    s.push(4)
    print(s.pop_at(0))
    print(s.pop())
    print(s.pop())
    print(s.pop())


if __name__ == "__main__":
    main()
